### ISSUE TITLE
AmazonDynamoDBClientBuilder takes ~2 seconds to build AmazonDynamoDB instance

### ISSUE THREAD

**Issue by stuartleylandcole** [aws/aws-sdk-java#1774]
I'm using the AWS Java SDK in a AWS Lambda function that connects to an AWS DynamoDB instance. I've noticed that the `AmazonDynamoDBClientBuilder.build()` method takes almost 2 seconds to build an `AmazonDynamoDB` instance the first time my Lambda is called. I use [Dagger](https://github.com/google/dagger) to ensure I have only one instance of the `AmazonDynamoDB` client so subsequent requests do not incur this cost but I'm trying to reduce the time taken for the initial startup.

This is the code that builds the instance:

```
    public AmazonDynamoDB amazonDynamoDb(final PropertiesService properties) {
        LOGGER.debug("Creating AmazonDynamoDB instance");

        final AmazonDynamoDBClientBuilder builder = AmazonDynamoDBClientBuilder.standard();

        final String awsRegion = properties.getAwsRegion();
        final String endpoint = properties.getDynamoEndpoint();
        if ((endpoint != null) && !endpoint.isEmpty()) {
            builder.withEndpointConfiguration(new EndpointConfiguration(endpoint, awsRegion));
        } else {
            builder.withRegion(awsRegion);

        }

        final AmazonDynamoDB dynamoDB = builder.build();
        LOGGER.debug("Finished creating AmazonDynamoDB");

        return dynamoDB;
    }
```

And these are some timings for the logging:

```
2018-10-02 09:51:16.332 DEBUG DaoModule:29 - Creating AmazonDynamoDB instance
2018-10-02 09:51:18.003 DEBUG DaoModule:43 - Finished creating AmazonDynamoDB
```

I have stepped through the `build()` method but haven't found anything obviously incriminating that could explain the slowness - is this roughly the amount of time that I should expect to create this object? If not is there anything I can do to speed it up?

**Relevant information**

- AWS SDK version: 1.11.412
- Source code:
  - [Repository](https://github.com/meal-plannr/aws-api)
  - [Specific class](https://github.com/meal-plannr/aws-api/blob/master/src/main/java/com/mealplanner/config/DaoModule.java#L28)

Thanks for any help you can offer 😃

**Comment by zoewangg** [comment]
Hi, @stuartleylandcole, v1 lambda cold start is an known issue and unfortunately, there is really not much we can do for v1 due to backwards compatibility reasons. One of the reasons why it takes such long time to just initialize the client is because loading ApacheHttpClient is very expensive. Initializing Jackson and log4j can also take a long time. One thing you can try to improve it a bit is to specify a credentials provider to avoid credentials lookup.

Good new is that we have made several changes in [aws-sdk-javav2](https://github.com/aws/aws-sdk-java-v2/) (most of the modules are currently in developer preview) to improve lambda cold start and it is now much faster in v2. More details: https://github.com/aws/aws-sdk-java-v2/issues/6

**Comment by zoewangg** [comment]
`HttpUrlConnection` should be fairly fast and you don't need to write code to adapt it to `SdkHttpClient` since it's part of the SDK in v2!  [url-connection-client](https://github.com/aws/aws-sdk-java-v2/blob/master/http-clients/url-connection-client/pom.xml)

>  Using Serverless framework and Lambda I'm assuming I would use EnvironmentVariableCredentialsProvider 

Yes, `EnvrionmentVariableCredentialsProvider` should work.

> When I was stepping through the code earlier it felt (although this is a little unscientific) that creating the ClientConfiguration was slow - do you have any tips that could help here?

Hmm, I'm not sure if `ClientConfiguration` could be an issue, but you can try to explicitly pass a `new ClientConfiguraion()` to avoid resolving to default one.

>  is there some way I can get insight into what's taking time when starting my Lambda? 

You can probably use some profiler tool to get insight on where the time is going.

Please try out v2 and let us know if you have any feedback! 😄

**Comment by stuartleylandcole** [comment]
@zoewangg reporting back with some progress...

SDK v2 has taken about 0.5-1s off the start up time which is great news; I'm now around 1.0-1.2s vs 1.7-2.0s with v1. I think this is mostly attributable to using `UrlConnectionHttpClient`. Specifying the credentials provider hasn't made a difference at this moment but I've logged an issue for that (see below)

I've logged https://github.com/aws/aws-sdk-java-v2/issues/748 and https://github.com/aws/aws-sdk-java-v2/issues/749 which seem to be the areas where most time is spent starting in v2 - the team seem to already be aware of these though and are planning to address them before GA.

I'm happy to close this issue now unless you think there's value in leaving it open? Thanks very much for your help and great work on SDK v2!
